File: 0120-triangle/0120-triangle.js
/**
 * @param {number[][]} triangle
 * @return {number}
 */
let matrix;
let dp;

function f_bu(){
    dp = Array(205);

    for(let i = 0; i < 205; i++){
        dp[i] = Array(205);
    }

    let n = matrix.length // total rows
    //base case
    for(let i = 0; i < matrix[n-1].length; i++){
        dp[n-1][i] = matrix[n-1][i];
    }

    for(let row = n-2; row >= 0; row--){
        for(let col = 0; col < matrix[row].length; col++){
            dp[row][col] = matrix[row][col] + Math.min( dp[row+1][col], dp[row+1][col+1]);
        }
    }

    return dp[0][0];
}


var minimumTotal = function(triangle) {
    matrix = triangle;

    return f_bu();
};

File: 0120-triangle/0120-triangle.java
class Solution {
    public static List<List<Integer>> matrix;
    public static int[][] dp = new int[205][205];

    public static int f(int row, int col){
        if(row == matrix.size() - 1) return matrix.get(row).get(col);

        if(dp[row][col] != -1) return dp[row][col];

        return dp[row][col] = matrix.get(row).get(col) + Math.min( f(row+1, col), f(row+1, col+1) );
    }
    public int minimumTotal(List<List<Integer>> triangle) {
        matrix = triangle;

        for(int i = 0; i < 205; i++){
            Arrays.fill(dp[i], -1);
        }

        return f(0,0);
    }
}